File: generated-code/scala/codecraft/TcpReadWrite/src/main/scala/trans_gen_test/model/Action.scala
package trans_gen_test.model

import trans_gen_test.util.StreamUtil

/**
 * Player's action
 *
 * @param entityActions New actions for entities. If entity does not get new action, if will continue to perform previously set one
 */
case class Action(entityActions: Map[Int, trans_gen_test.model.EntityAction]) {
    /**
     * Write Action to output stream
     */
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, entityActions.size)
        entityActions.foreach { case (key, value) =>
            StreamUtil.writeInt(stream, key)
            value.writeTo(stream)
        }
    }

    /**
     * Get string representation of Action
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("Action { ")
        stringBuilder.append("entityActions: ")
        stringBuilder.append(entityActions)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object Action {
    /**
     * Read Action from input stream
     */
    def readFrom(stream: java.io.InputStream): Action = Action(
        (0 until StreamUtil.readInt(stream)).map { _ => (
            StreamUtil.readInt(stream),
            trans_gen_test.model.EntityAction.readFrom(stream)
        )}.toMap
    )
}

File: generated-code/kotlin/codecraft/TcpReadWrite/src/main/kotlin/trans_gen_test/model/Action.kt
package trans_gen_test.model

import trans_gen_test.util.StreamUtil

/**
 * Player's action
 */
class Action {
    /**
     * New actions for entities. If entity does not get new action, if will continue to perform previously set one
     */
    var entityActions: MutableMap<Int, trans_gen_test.model.EntityAction>

    constructor(entityActions: MutableMap<Int, trans_gen_test.model.EntityAction>) {
        this.entityActions = entityActions
    }

    /**
     * Write Action to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, entityActions.size)
        for (entityActionsEntry in entityActions) {
            val entityActionsKey = entityActionsEntry.key
            StreamUtil.writeInt(stream, entityActionsKey)
            val entityActionsValue = entityActionsEntry.value
            entityActionsValue.writeTo(stream)
        }
    }

    /**
     * Get string representation of Action
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("Action { ")
        stringBuilder.append("entityActions: ")
        stringBuilder.append(entityActions)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read Action from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Action {
            var entityActions: MutableMap<Int, trans_gen_test.model.EntityAction>
            val entityActionsSize = StreamUtil.readInt(stream)
            entityActions = mutableMapOf();
            for (entityActionsIndex in 0 until entityActionsSize) {
                var entityActionsKey: Int
                entityActionsKey = StreamUtil.readInt(stream)
                var entityActionsValue: trans_gen_test.model.EntityAction
                entityActionsValue = trans_gen_test.model.EntityAction.readFrom(stream)
                entityActions.put(entityActionsKey, entityActionsValue)
            }
            return Action(entityActions)
        }
    }
}